### PR TITLE
fix(v1): eval writes summary.json beside traces.jsonl

### PR DESCRIPTION
**Pull request by willccbb**
## What the run dir lacked

A finished `eval` run dir held `traces.jsonl`, `configs/resolved/eval.json` and `logs/`, and nothing that said how the run went. The dashboard computed a mean reward and an error share for display only, so anyone reading a run afterwards (a script, a pipeline, a person with `jq`) re-derived pass rate, mean reward and failures from `traces.jsonl` with their own reading of `Episode.ok`, `Trace.reward` and `agent.trainable`. Those readings drift. The constraint here is one reading of the record, owned by verifiers.

## What changes

- `cli/output.py` gains `summarize(episodes) -> Summary` and `write_summary`. `run_eval` writes `summary.json` beside `traces.jsonl` when the run completes and logs one `summary: episodes=… failed=… reward=…` line (on the console under `--no-rich`, in `logs/latest/eval.log` always). A `--resume` drops the previous summary before re-running the owed rollouts, since `resume.load` rewrites `traces.jsonl` underneath it, and writes a fresh one on completion. The file's presence therefore means the run completed.
- The dashboard footer's `reward … · err …` reads `summarize` instead of its own formula; the per-signal breakdown rows keep `format_mean`.
- `docs/v1/evaluation.md` and the `evaluate-environments` skill name the file.
- No new config, flag or dependency.

## What `summary.json` holds

```json
{
  "episodes": 6,
  "failed": 2,
  "errors": {"ProviderError": 2},
  "reward": 0.5,
  "tasks": {
    "echo:hello world": {"name": null, "rollouts": 2, "failed": 0, "reward": 1.0},
    "echo:ping":        {"name": null, "rollouts": 2, "failed": 2, "reward": null},
    "echo:verifiers":   {"name": null, "rollouts": 2, "failed": 0, "reward": 0.0}
  }
}
```

- `episodes`: finished episodes, one per rollout (on a resume, the kept ones plus the re-run ones).
- `failed`: episodes that ended not ok (`Episode.ok` is false).
- `errors`: failed episodes by the type of the error that failed them: a failed trace's last error, else the episode's own last error (a hook's). These are the live errors `episode_should_retry` reads; an ok trace's errors are history its per-agent retry recovered from, and are never the cause. A failed episode that recorded no error is counted nowhere here, so the counts sum to at most `failed`.
- `reward`: mean over scored rollouts. A rollout's reward is the mean `Trace.reward` over its policy traces (`agent.trainable`, or every trace when no seat is trainable) that were scored; a trace whose rewards are all `None` (or empty) was never scored, and a rollout with no scored trace is unscored and outside every mean. `null` when nothing was scored.
- `tasks`: the same per task, keyed by `task.key` (`task.hash` when unset; rows written before either was recorded hash their data, as resume does), with `TaskData.name` when the task has one.

`utils/platform.run_metrics` (the `--push` metadata) keeps its v0-shaped `avg_reward` / `avg_error`, where an errored trace counts as 0. That is a platform contract and is untouched here.

## How it was checked

- `tests/v1/test_e2e.py::test_summary` runs `echo-v1` on the null harness (subprocess runtime, in-process) against a local stub endpoint that answers every request with "hello world" and fails the `fail` phrase upstream, so it needs no key and no network. It asserts `summary.json` verbatim: 3 episodes, 1 failed, `{"ProviderError": 1}`, reward 0.5, per task 1.0 / 0.0 / null. `test_summary_rules` pins the rules that run can't reach over hand-built episodes: a frozen seat's reward is not counted, every trace counts when no seat is trainable, an unscored trace leaves its rollout out of the mean, and a failure is attributed to the failed trace's error rather than to history a recovered sibling kept.
- The default served path through the CLI (`uv run eval echo-v1 --env.agent.harness.id null --env.agent.runtime.type subprocess --no-rich -n 3 -r 2`, same stub): the run dir gains the `summary.json` above, the log ends with `summary: episodes=6 failed=2 reward=0.500`, and a `--resume` of that dir re-runs the two failed rollouts and rewrites the summary.
- `uv run ruff check --fix .`, `uv run ruff format .`, `uv run pytest tests/v1 -n auto -m "not e2e"` (85 passed), `uv run pre-commit run --all-files` and `ty check verifiers` are green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Additive run output and aligned dashboard display; resume behavior only removes a stale summary file before recomputation.
> 
> **Overview**
> Finished **eval** runs now emit **`summary.json`** next to **`traces.jsonl`**, giving pipelines and humans a single canonical rollup instead of re-parsing episodes with ad hoc reward/error rules.
> 
> **`summarize(episodes)`** in `cli/output.py` defines that rollup: episode and failure counts, failures grouped by error type, mean reward over **scored** rollouts (policy/`trainable` traces only, with documented fallbacks for multi-agent and unscored cases), plus the same stats per task key. **`run_eval`** writes the file on completion, logs a one-line summary, and **deletes the old `summary.json` on `--resume`** before rewriting it when the resumed run finishes. The live dashboard **reward · err** footer now uses **`summarize`** so on-screen numbers match the file.
> 
> Docs/skills mention **`summary.json`**. Tests add a local **`stub_model`** HTTP stub (no API key/network), an e2e check of the echo run’s summary shape, and unit tests for aggregation edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2331dcd70248e2cca06b87145639fd45e423892d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Write `summary.json` beside `traces.jsonl` on eval completion
> - Adds `Summary` and `TaskSummary` Pydantic models plus `summarize` and `write_summary` helpers in [output.py](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-07fce45dd0fa633f34d25e19962ab801b705a509774833de0d81bfe872787e8b) to aggregate episode counts, failure counts by error type, scored reward means, and per-task results
> - Integrates summary generation into [runner.py](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-65be4a576330571f2729a3bbcdbe92ac81bbcba450c313597c1bdca3a1c3f2e4): `run_eval` writes `summary.json` after all episodes finish and deletes any stale summary before a resumed run's owed rollouts
> - Updates the live dashboard in [eval.py](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-0a27b905958929f3ae9e01937dd7e3cf97a60dec137caddcbf87792784fc9035) so headline reward and error values follow the `Summary` aggregation rules (reward mean excludes unscored rollouts; error ratio uses failed/total episodes)\n- Adds a local HTTP stub-model fixture and end-to-end/aggregation tests in [test_e2e.py](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-ebbbd919487889487f891b359a756e6c4630e106b16ddaef8ccddc6b900a96e1) and [conftest.py](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-bc28790d38dc6e7e4f2c44639d40bc33dea724d6c73149abf7ba82b37d14f40b), plus docs in [evaluation.md](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-dcded4c488b9e44f4750d10743f8f2319c8ff959fae7ca1a23d3d81c88a64848) and [SKILL.md](https://github.com/PrimeIntellect-ai/verifiers/pull/2520/files#diff-36d08ee11d380ecaa5beacd5b98cd8bfa0791dfa784b93956b128cafc58dbe0d)
> - Behavioral Change: resumed evaluations now remove and rewrite `summary.json`; reward means prefer trainable-agent traces and exclude unscored rollouts, which may change displayed dashboard values versus prior direct calculation
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 2331dcd.</sup>
> 
>

### CODE REVIEW

File: verifiers/v1/cli/output.py [hunk 1]
```diff
@@ -203,3 +209,90 @@ async def append_trace(
...
+        },
+    )
+
+
+def write_summary(results_dir: Path, summary: Summary) -> Path:
+    """Write the completed run's `summary.json` beside its traces; return its path. Nulls
+    included: an unscored reward reads as `null`, never as a missing key."""
+    path = results_dir / SUMMARY_FILE
+    path.write_text(summary.model_dump_json(indent=2))
+    return path
```
[the remark this answers is not in the document]
Comment (willccbb) [reply verifiers/v1/cli/output.py:300]: `summary.model_dump_json` runs before `write_text` opens the file, so a serialization failure leaves no file behind, and the summary is written only after every episode has landed in `traces.jsonl`, so a torn file cannot mark an incomplete run as complete; it would be a malformed record of a completed run, which a reader fails loudly on. `write_config` in this module writes its JSON the same way; the temp-file swap is reserved for rewriting `traces.jsonl` and the `logs/latest` symlink, where a torn write loses existing data. Holding as is.

File: verifiers/v1/cli/output.py [hunk 2]
```diff
@@ -203,3 +209,92 @@ async def append_trace(
...
+            for key, group in by_task.items()
+        },
+    )
+
+
+def write_summary(results_dir: Path, summary: Summary) -> Path:
+    """Write the completed run's `summary.json` beside its traces; return its path. Nulls
+    included: an unscored reward reads as `null`, never as a missing key."""
+    path = results_dir / SUMMARY_FILE
+    path.write_text(summary.model_dump_json(indent=2))
```
Comment (xeophon) [inline verifiers/v1/cli/output.py:299]: `summary.json` is the run-completion artifact, but `Path.write_text()` creates or truncates the final path before the write completes. An interruption or full-disk write can leave malformed JSON that a watcher observes as completed, and `--resume` will not repair it when no rollouts are owed. Please serialize first, write a sibling temporary file, and atomically replace the final path.

File: verifiers/v1/cli/eval/runner.py [hunk 3]
```diff
@@ -176,6 +179,9 @@ async def run_eval(config: EvalConfig) -> list[Episode]:
             raise SystemExit(0)
         counts = distribute(keys, owed, config.num_rollouts)
         plan = [(task, n) for task, n in zip(tasks, counts) if n]
+        # The kept rows are a new run state: the summary described the old one, and
+        # the run rewrites it on completion.
+        (out / SUMMARY_FILE).unlink(missing_ok=True)
```
Comment (xeophon) [inline verifiers/v1/cli/eval/runner.py:184]: `resume.load()` above atomically rewrites `traces.jsonl` to the kept rows before this removes the old summary. If the process stops between those operations, a stale `summary.json` remains beside a changed trace set and still looks like a completion marker. Please invalidate the summary before calling `resume.load()`, while preserving or republishing a valid summary for the no-owed path.
